core/src/test/scala/unit/kafka/log/LogCleanerLagIntegrationTest.scala

```diff
@@ -0,0 +1,192 @@
+
+  val compactionLag = 1 * msPerHour
+  assertTrue("compactionLag must be divisible by 2 for this test", compactionLag % 2 == 0)
+
+  val time = new MockTime(1400000000000L)  // Tue May 13 16:53:20 UTC 2014
+  val cleanerBackOffMs = 200L
+  val segmentSize = 100
+  val deleteDelay = 1000
+  val logName = "log"
+  val logDir = TestUtils.tempDir()
+  var counter = 0
+  val topics = Array(TopicAndPartition("log", 0), TopicAndPartition("log", 1), TopicAndPartition("log", 2))
+//  val compressionCodec = CompressionCodec.getCompressionCodec("none")
+  val compressionCodec = CompressionCodec.getCompressionCodec(compressionCodecName)
+
+  @Test
+  def cleanerTest() {
+    val cleaner = makeCleaner(parts = 3, backOffMs = cleanerBackOffMs)
+    val log = cleaner.logs.get(topics(0))
+
+    // t = T0
+    val T0 = time.milliseconds
+    val appends0 = writeDups(numKeys = 100, numDups = 3, log, compressionCodec, timestamp = T0)
+    val startSizeBlock0 = log.size
+
+    // force the modification time of the log segments
+    log.logSegments.foreach(_.log.file.setLastModified(T0))
+    log.logSegments.foreach { s => assertEquals("Segments should have lastModified time of T0", T0, s.lastModified) }
+
+    val activeSegAtT0 = log.activeSegment
+    debug(s"active segment at T0 has base offset: ${activeSegAtT0.baseOffset}")
+
+    cleaner.startup()
+
+    // T0 < t < T1
+    // advance to a time still less than one compaction lag from start
+    time.sleep(compactionLag/2)
+    Thread.sleep(5 * cleanerBackOffMs) // give cleaning thread a chance to _not_ clean
+    assertEquals("There should be no cleaning until the compaction lag has passed", startSizeBlock0, log.size)
+
+    // t = T1 > T0 + compactionLag
+    // advance to time a bit more than one compaction lag from start
+    time.sleep(compactionLag/2 + 1)
+    val T1 = time.milliseconds
+
+    // write another block of data
+    val appends1 = appends0 ++ writeDups(numKeys = 100, numDups = 3, log, compressionCodec, timestamp = T1)
+    val firstBlock1SegmentBaseOffset = activeSegAtT0.baseOffset
+    log.logSegments.foreach { s =>
+      if (s.baseOffset >= activeSegAtT0.baseOffset)
+        s.log.file.setLastModified(T1)
+    }
+
+    // the first block should get cleaned
+    cleaner.awaitCleaned("log", 0, activeSegAtT0.baseOffset)
+
+    // check the data is the same
+    val read1 = readFromLog(log)
+    assertEquals("Contents of the map shouldn't change.", appends1.toMap, read1.toMap)
+
+    val compactedSize = log.logSegments(0L, activeSegAtT0.baseOffset).map(_.size).sum
+    val lastCleaned = cleaner.cleanerManager.allCleanerCheckpoints.get(TopicAndPartition("log", 0)).get
+    assertTrue(s"log cleaner should have processed up to offset $firstBlock1SegmentBaseOffset, but lastCleaned=$lastCleaned", lastCleaned >= firstBlock1SegmentBaseOffset)
+    assertTrue(s"log should have been compacted:  startSize=$startSizeBlock0 compactedSize=$compactedSize", startSizeBlock0 > compactedSize)
```

Comment:
To be more accurate, it seems that startSizeBlock0 should be computed as log.logSegments(0L, activeSegAtT0.baseOffset).map(_.size).sum in line 66?